org/mozilla/jss/pkcs11/PK11Signature.c

```diff
@@ -355,6 +454,81 @@ getAlgorithm(JNIEnv *env, jobject sig)
     return retval;
 }
 
+static SECOidTag
+getDigestAlgorithm(JNIEnv *env, jobject sig)
+{
+    jclass sigClass;
+    jfieldID algField;
+    jobject alg;
+    SECOidTag retval=SEC_OID_UNKNOWN;
+
+    PR_ASSERT(env!=NULL && sig!=NULL);
+
+    sigClass = (*env)->GetObjectClass(env, sig);
+    PR_ASSERT(sigClass != NULL);
+
+    algField = (*env)->GetFieldID(  env,
+                                    sigClass,
+                                    SIG_DIGEST_ALGORITHM_FIELD,
+                                    SIG_ALGORITHM_SIG);
+    if(algField == NULL) {
+        ASSERT_OUTOFMEM(env);
+        goto finish;
+    }
+
+    alg = (*env)->GetObjectField(env, sig, algField);
+    if(alg == NULL) {
+        ASSERT_OUTOFMEM(env);
+        goto finish;
+    }
+
+    retval = JSS_getOidTagFromAlg(env, alg);
+
+finish:
+    return retval;
+}
+
+static
+
+SECStatus getRSAPSSParamsAndSigningAlg(JNIEnv *env, jobject this, PRArenaPool *arena, SECAlgorithmID **alg, SECKEYPrivateKey *privk)
+{
+    SECItem *sigAlgParams;
+    SECAlgorithmID *signAlg=NULL;
+    SECOidTag digestAlg = SEC_OID_UNKNOWN;
+    SECStatus rv = SECSuccess ;
```

Comment:
Oh here's where that space went. ;-) I'd remove this one. 

```c
SECStatus rv = SECSuccess;
```

Also, do we want this to be `SECFailure`?

I'm confused:

```c
if (alg == NULL) {
    return SECSuccess;
}
```

Why would we want a success exit code here? 

```c
    if (signAlg == NULL) {
        return SECSuccess;
    }
```

Also seems like the wrong return if `PORT_AreanZAlloc` fails.

```c

    if (!sigAlgParams) {
        rv = SECFailure;
        goto finish;
    }
```

That seems right. :-) But we can just initialize `rv` to `SECSuccess` and remove the set here. :)

Reply:
Actually, going a step further, I don't think the `finish` label is really required. The `arena` handles freeing the memory. So we can replace `goto finish` with `return rv` directly.

Reply:
Apologies, was still working on this latish last night :)

Reply:
Fixed this to the spirit of what you were suggesting...